lib/recursive-methods.rb

```diff
@@ -1,49 +1,106 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError.new("Negative numbers aren't allowed.")
+    else
+        n == 0 ? 1:n * factorial(n-1)
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)?
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    else
+        rs = reverse(s[1..-1])
+        rs += s[0]
+        rs
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    else
+        rs = reverse(s[1..-1])
+        rs += s[0]
+        rs
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
-end
+    if n == 0
+        return 0
+    else
+        return 2 + bunny(n-1)
+    end 
+end 
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    b = 0
+    e = s.length - 1
+
+    def valid?(current_beg, current_end, s)
+        if current_beg > current_end 
+            return true 
+        elsif s[current_beg] == '(' && s[current_end] == ')'
+            valid?(current_beg + 1, current_end - 1, s)
+        else
+            return false
+        end 
+    end 
+    return valid?(b,e,s)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def search(array, value, b = 0)
+    if b >= array.length
+        return false
+    elsif array[b] == value
+        return true
+    else
+        return search(array, value, b + 1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log10n)
+# Space complexity: O(n)
 def is_palindrome(s)
```

Comment:
How would this be a log in time complexity?  Were you thinking about the digit_match problem?
This is O(n) in time complexity.

Because you're creating a new array with each recursive call, this is O(n^2) in space complexity.